src/replica/AsyncTimer.cc

```diff
@@ -0,0 +1,60 @@
+/*
+ * LSST Data Management System
+ *
+ * This product includes software developed by the
+ * LSST Project (http://www.lsst.org/).
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the LSST License Statement and
+ * the GNU General Public License along with this program.  If not,
+ * see <http://www.lsstcorp.org/LegalNotices/>.
+ */
+
+// Class header
+#include "replica/AsyncTimer.h"
+
+using namespace std;
+
+namespace lsst::qserv::replica {
+
+AsyncTimer::AsyncTimer(boost::asio::io_service& io_service, unsigned int expirationIvalMs,
+                       CallbackType const& onFinish)
+        : _io_service(io_service),
+          _expirationIvalMs(expirationIvalMs),
+          _onFinish(onFinish),
+          _timer(_io_service) {}
+
+AsyncTimer::~AsyncTimer() { _timer.cancel(); }
```

Comment:
For the exceptions, you're right. I should switch to using the non-exception version of the method:
```
boost::system::error_code ec;
_timer.cancel(ec);
```
And ignore the code ... or just report it via the LSST Logger.

For the second comment, I agree this may be a potential problem in a general case, and it has to be addressed to avoid running into problems with the life expectancy of this object. A solution here would be twofold:
- deriving this class from `std::enable_shared_from_this<AsyncTimer>`
- capturing a copy of the shared pointer in the callback handler of the time
- and resetting the user-provided callback to `nullptr` before canceling the timer. The timer's callback handler is already checking if the user calback wasn't reset.
This is the same technique that is used in the rest of the Replication Framework.